tools/util/ReplayInterpreter.cpp

```diff
@@ -0,0 +1,716 @@
+        std::cerr << "Warning, version mismatch:\n"
+          << "  Tool version: " << UMPIRE_VERSION_MAJOR << "." << UMPIRE_VERSION_MINOR << "." << UMPIRE_VERSION_PATCH << std::endl
+          << "  Log  version: "
+          << m_json["payload"]["major"] << "."
+          << m_json["payload"]["minor"]  << "."
+          << m_json["payload"]["patch"]  << std::endl;
+      }
+    }
+    else {
+      std::cerr << "Unknown Replay (" << m_json["event"] << ")\n";
+      return -1;
+    }
+
+    compare_ss << std::endl;
+    raw_line = m_line;
+    sym_line = compare_ss.str();
+    return 1;
+  }
+  return 0;   // EOF
+}
+
+ReplayInterpreter::ReplayInterpreter( std::string in_file_name ):
+    m_input_file(in_file_name), m_num_allocators(0), m_op_seq(0)
+{
+  if ( ! m_input_file.is_open() ) {
+    std::cerr << "Unable to open input file " << in_file_name << std::endl;
+    exit (1);
+  }
+}
+
+template <typename T> void get_from_string( const std::string& s, T& val )
+{
+    std::istringstream ss(s);
+    ss >> val;
+}
+
+void ReplayInterpreter::strip_off_base(std::string& s)
+{
+  const std::string base("_base");
+
+  if (s.length() > base.length()) {
+    if (s.compare(s.length() - base.length(), base.length(), base) == 0) {
+      s.erase(s.length() - base.length(), base.length());
+    }
+  }
+}
+
+void ReplayInterpreter::replay_makeAllocator( void )
+{
+  const std::string& allocator_name = m_json["payload"]["allocator_name"];
+
+  //
+  // When the result isn't set, just perform the operation.  We will
+  // establish the mapping on after the result has been recorded in the
+  // two-step REPLAY process for this event.
+  //
+  if ( m_json["result"].is_null() ) {
+    const bool introspection = m_json["payload"]["with_introspection"];
+    const std::string type = m_json["payload"]["type"];
+
+    if ( type == "umpire::strategy::AllocationAdvisor" ) {
+      const int numargs = static_cast<int>(m_json["payload"]["args"].size());
+      const std::string& base_allocator_name = m_json["payload"]["args"][0];
+      const std::string& advice_operation = m_json["payload"]["args"][1];
+      const std::string& last_arg = m_json["payload"]["args"][numargs-1];
+
+      //
+      // The last argument to this constructor will either be a string or
+      // will be an integer.  If it is an integer, we assume that it is the
+      // optional device_id argument.
+      //
+      int device_id = -1;   // Use default argument if negative
+      if (last_arg.find_first_not_of( "0123456789" ) == std::string::npos) {
+        std::stringstream ss(last_arg);
+        ss >> device_id;
+      }
+
+      if (device_id >= 0) { // Optional device ID specified
+        switch ( numargs ) {
+        default:
+          std::cerr << "Invalid number of arguments (" << numargs
+            << " for " << type << " operation.  Stopping" << std::endl;
+          exit(1);
+        case 3:
+          compare_ss << introspection 
```

Comment:
I think you can save many lines of code at the higher level by building this inside `makeAdvisor(), makeDynamicPool()`, etc.

Reply:
Perhaps.  If possible, I will do this under a separate PR